### PR TITLE
Enable MGTwoLevelTransfer::reinit_geometric_transfer() for refinement levels

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

**Comment by peterrum** [comment]
I have updated this PR. Local smoothing will not work `MGTransferGlobalCoarsening`, since it would need a more global view and only two-level informations to be able to transfer the solution from and to the multigrid levels. This PR at least allows to run the algorithm on refinement levels without needing to create new `Triangulation` objects.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 1]
```diff
@@ -865,28 +878,46 @@ namespace internal
...
+              if (!cell->is_artificial() && cell->is_locally_owned())
+                op(cell);
           }
+        else
+          {
+            for (const auto &cell : tria.cell_iterators_on_level(level))
+              if (cell->level_subdomain_id() == tria.locally_owned_subdomain())
+                op(cell);
+          }
+      };
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:917]: I do not really like the nested use of several lambda functions here as it makes it somewhat complicated to see what is going on, but I admit that I don't see a simple solution without code duplication or collection of the cells to loop over in some different way, so let's keep it like this.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 2]
```diff
@@ -1852,7 +1985,7 @@ namespace internal
         std::vector<unsigned int *> level_dof_indices_fine_(
           fe_index_pairs.size());
 
-        for (unsigned int i = 0; i < fe_index_pairs.size(); ++i)
+        for (unsigned int i = 0; i < fe_index_pairs.size(); i++)
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:1988]: Can you switch this one back?

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 3]
```diff
@@ -2059,7 +2192,7 @@ namespace internal
             fe_index_pairs.size());
           std::vector<Number *> weights_(fe_index_pairs.size());
 
-          for (unsigned int i = 0; i < fe_index_pairs.size(); ++i)
+          for (unsigned int i = 0; i < fe_index_pairs.size(); i++)
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:2195]: same here

**Review by kronbichler — APPROVED** [review]

**Comment by peterrum** [comment]
@kronbichler I have done the changes!

**Review by kronbichler — APPROVED** [review]

**Pull request merged**